lib/bookmark_query.rb

```diff
@@ -27,54 +27,135 @@ def initialize(user:, guardian: nil, params: {})
-    results = results.merge(topics.or(pms))
 
-    results = results.merge(Post.secured(@guardian))
+    # A note about the difference in queries here...pre-polymorphic bookmarks are
+    # all attached to a post so both the Post.secured filter and the topics/pms/allowed
+    # category filters all work correctly. However with polymorphic bookmarks, the
+    # bookmarks could be attached to any relation, so we must get the post and
+    # topic bookmarks separately and apply the relevant filters to them directly.
+    #
+    # Much of the complexity in this file will be cleaned up when we switch completely
+    # to polymorphic bookmakrks.
+    if SiteSetting.use_polymorphic_bookmarks
+      results = list_all_results_polymorphic(topics, pms)
+    else
+      results = list_all_results(topics, pms)
+    end
+
+    results = results.order(
+      "(CASE WHEN bookmarks.pinned THEN 0 ELSE 1 END),
+        bookmarks.reminder_at ASC,
+        bookmarks.updated_at DESC"
+    )
 
     if @params[:q].present?
-      term = @params[:q]
-      bookmark_ts_query = Search.ts_query(term: term)
-      results = results
-        .joins("LEFT JOIN post_search_data ON post_search_data.post_id = bookmarks.post_id")
-        .where(
-          "bookmarks.name ILIKE :q OR #{bookmark_ts_query} @@ post_search_data.search_data",
-          q: "%#{term}%"
-        )
+      if SiteSetting.use_polymorphic_bookmarks
+        results = polymorphic_search(results, @params[:q])
+      else
+        results = search(results, @params[:q])
+      end
     end
 
     if @page.positive?
       results = results.offset(@page * @params[:per_page])
     end
 
     results = results.limit(@limit)
+    BookmarkQuery.preload(results, self)
+    results
+  end
+
+  private
+
+  def list_all_results(topics, pms)
+    results = base_bookmarks.merge(topics.or(pms))
+    results = results.merge(Post.secured(@guardian))
+    results = @guardian.filter_allowed_categories(results)
 
     if BookmarkQuery.preloaded_custom_fields.any?
-      Topic.preload_custom_fields(
-        results.map(&:topic), BookmarkQuery.preloaded_custom_fields
-      )
+      Topic.preload_custom_fields(results, BookmarkQuery.preloaded_custom_fields)
     end
 
-    BookmarkQuery.preload(results, self)
-
-    @guardian.filter_allowed_categories(results)
+    results
   end
 
-  private
+  def list_all_results_polymorphic(topics, pms)
+    topic_results = topic_bookmarks.merge(topics.or(pms))
+
+    post_results = post_bookmarks.merge(topics.or(pms))
+    post_results = post_results.merge(Post.secured(@guardian))
+
+    topic_results = @guardian.filter_allowed_categories(topic_results)
+    post_results = @guardian.filter_allowed_categories(post_results)
+
+    if BookmarkQuery.preloaded_custom_fields.any?
+      Topic.preload_custom_fields(topic_results, BookmarkQuery.preloaded_custom_fields)
+      Topic.preload_custom_fields(post_results, BookmarkQuery.preloaded_custom_fields)
```

Comment:
I don't think works. `Topic.preload_custom_fields` is expecting an array of `Topic` objects not a AR relation.

Reply:
Also, I don't think preloading here is going to be useful when the query below would be creating new AR objects.

Reply:
Hmmm I am going to have to think about this, the preloaded custom fields are only relevant for post and topic bookmarks (I think) at this time. Will look at what is using these bookmark query preloaded fields.